### DISCUSSION TITLE
How to add new node type to the AST for `[[_TOC_]]` and `[[_TOSP_]]` ?

### DISCUSSION THREAD

**Discussion by doggy8088** [remarkjs/remark#1258] [Q&A] [answered]
I'd like to implement a plugin that support custom syntax offered by [Wiki page TOC of the Azure DevOps Services](https://learn.microsoft.com/en-us/azure/devops/project/wiki/wiki-markdown-guidance?view=azure-devops&WT.mc_id=DT-MVP-4015686#table-of-contents-toc-for-wiki-pages).

The syntax will be like this:

```txt
[[_TOC_]]

[[_TOSP_]]
```

Here is the code I wrote currently:

```ts
let parsed = unified()
  .use(remarkParse)
  .use(remarkGfm)
  .use(remarkFrontmatter, ['yaml', 'toml'])
  .use(remarkStringify, {
    emphasis: '_',
    rule: '-',
  })
  .processSync(txt).toString();
```

After running my code, the Markdown will become this:

```md
\[\[_TOC_]]

\[\[_TOSP_]]
```

But I want the result is:

```md
[[_TOC_]]

[[_TOSP_]]
```

Is that possible?

I need some help on how to implement this. Is there any docs or reference for me? Thanks.

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @doggy8088! :wave:

> How to add new node type to the AST for [[_TOC_]] and [[_TOSP_]] ? 

It is possible, though generally not recommended to create yet another markdown flavor (https://github.com/micromark/micromark#extending-markdown)
If you do want to add token type/AST node, see https://github.com/micromark/micromark#creating-a-micromark-extension
And draw some inspiration from https://github.com/landakram/remark-wiki-link

---

An alternative would be nudging Azure/Microsoft toward using standard markdown directives https://github.com/remarkjs/remark-directive#readme, instead of creating another markdown flavor.

**Reply by doggy8088** [reply]
Thank you for the quick and kind response! 👍 

Is there any way to make some `text` node been avoid escaped? I don't want to create another markdown flavor either. It because my original requirement is simply keep `[[_TOC_]]` or `[[_TOSP_]]` been untouched (not-escaped).

**Reply by ChristianMurphy** [reply]
> Is there any way to make some text node been avoid escaped?

No, see https://github.com/orgs/remarkjs/discussions/1067, https://github.com/remarkjs/remark/issues/1168, https://github.com/remarkjs/react-markdown/issues/726, https://github.com/remarkjs/remark/issues/1245

`remark` is an Abtract Syntax Tree, it does not track if the original content was escaped or not.
Escapes are added in some spots to avoid confusion around if a syntax feature will trigger or not.

Your example, the text is link reference-like, the escape is added to avoid the following confusion on whether the reference is defined or not.

---
example 1:

this will render a link, and will not render `[]` around the text.

```md
[TOC]

... pages of content can go here ...

[TOC]: http://example.com
```

example 2:

this will only render text (no link) and will show the `[]` before and after.

```md
[TOC]

... pages of content can go here ...
```
